File: algorithms/0002. Add Two Numbers.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution 
{
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) 
    {
        ListNode p = l1;
        ListNode q = l2;
        ListNode ans = null;
        ListNode c = null;
        
        int m = 0;
        
        while (p != null || q != null)
        {
            int x = p != null ? p.val : 0;
            int y = q != null ? q.val : 0;
            
            int sum = (x + y + m) % 10;
            m = (x + y + m) / 10;
            
            if (c == null)
            {
                c = new ListNode(sum);
                ans = c;
            }
            else
            {
                c.next = new ListNode(sum);
                c = c.next;
            }
            
            p = p?.next;
            q = q?.next;
        }
        
        if (m > 0)
        {
            c.next = new ListNode(m);
        }
        
        return ans;
    }
}

File: algorithms/0002. Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution 
{
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) 
    {
        ListNode* ans = NULL;
        ListNode* c = NULL;
        
        int m = 0;
        
        while (l1 != NULL || l2 != NULL)
        {
            int x = l1 != NULL ? l1->val : 0;
            int y = l2 != NULL ? l2->val : 0;
            
            int sum = (x + y + m) % 10;
            m = (x + y + m) / 10;
            
            if (c == NULL)
            {
                c = new ListNode(sum);
                ans = c;
            }
            else
            {
                c->next = new ListNode(sum);
                c = c->next;
            }
            
            l1 = l1 != NULL ? l1->next : NULL;
            l2 = l2 != NULL ? l2->next : NULL;
        }
        
        if (m > 0)
        {
            c->next = new ListNode(m);
        }
        
        return ans;
    }
};